{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7996b360",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "os.chdir(\"D:/functional-prediction/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f8719f83",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "372\n"
     ]
    }
   ],
   "source": [
    "df_variant = pd.read_csv(\"output/dataset_review.csv\", dtype=str).fillna(\"\")\n",
    "\n",
    "df_variant[\"function\"] = df_variant[\"function\"].str.lower().str.strip()\n",
    "\n",
    "df_variant = df_variant[(~df_variant[\"function\"].str.contains(\"uncertain\")) & (~df_variant[\"function\"].str.contains(\"unknown\"))]\n",
    "print(len(df_variant))\n",
    "df_variant.to_csv(\"output/dataset_clean.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "45e28b7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_multi = pd.read_csv(\"output/dataset_multi_var_review.csv\", dtype=str).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "49071bb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 区分snp和indel\n",
    "df_variant[\"variant\"] = [\"\"] * len(df_variant)\n",
    "\n",
    "df_all = pd.concat([df_variant, df_multi], axis=0)\n",
    "\n",
    "type_list = []\n",
    "for index, row in df_all.iterrows():\n",
    "    if row[\"reference_allele\"] == \"-\" or row[\"variant_allele\"] == \"-\" or \"del\" in row[\"variant_allele\"] or \\\n",
    "    \"del\" in row[\"variant\"] or \"ins\" in row[\"variant\"]:\n",
    "        type_list.append(\"indel\")\n",
    "    else:\n",
    "        type_list.append(\"snp\")\n",
    "df_all[\"type\"] = type_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "82383553",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "34\n"
     ]
    }
   ],
   "source": [
    "df_indel = df_all[df_all[\"type\"] == \"indel\"]\n",
    "print(len(df_indel))\n",
    "df_indel.to_csv(\"output/dataset_clean_indel.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9a426bb0",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_variant = df_all[df_all[\"type\"] == \"snp\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "c283a16d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>variant_start</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>variant_allele</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>19</td>\n",
       "      <td>38500898</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>135</th>\n",
       "      <td>19</td>\n",
       "      <td>38496283</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>124</th>\n",
       "      <td>19</td>\n",
       "      <td>38448712</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>19</td>\n",
       "      <td>38446710</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>213</th>\n",
       "      <td>6</td>\n",
       "      <td>18130687</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>X</td>\n",
       "      <td>154536019</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>X</td>\n",
       "      <td>154536019</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>348</th>\n",
       "      <td>X</td>\n",
       "      <td>154535270</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>X</td>\n",
       "      <td>154535270</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>326</th>\n",
       "      <td>X</td>\n",
       "      <td>154534389</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>327</th>\n",
       "      <td>X</td>\n",
       "      <td>154534389</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>X</td>\n",
       "      <td>154533605</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307</th>\n",
       "      <td>X</td>\n",
       "      <td>154533605</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>304</th>\n",
       "      <td>X</td>\n",
       "      <td>154533596</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>X</td>\n",
       "      <td>154533596</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>279</th>\n",
       "      <td>X</td>\n",
       "      <td>154532765</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>280</th>\n",
       "      <td>X</td>\n",
       "      <td>154532765</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>251</th>\n",
       "      <td>X</td>\n",
       "      <td>154532625</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>252</th>\n",
       "      <td>X</td>\n",
       "      <td>154532625</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>234</th>\n",
       "      <td>X</td>\n",
       "      <td>154532269</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>235</th>\n",
       "      <td>X</td>\n",
       "      <td>154532269</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>231</th>\n",
       "      <td>X</td>\n",
       "      <td>154532258</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>232</th>\n",
       "      <td>X</td>\n",
       "      <td>154532258</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>7</td>\n",
       "      <td>99672916</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>222</th>\n",
       "      <td>7</td>\n",
       "      <td>99665212</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>1</td>\n",
       "      <td>97883368</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>1</td>\n",
       "      <td>97883353</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>1</td>\n",
       "      <td>97883352</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>1</td>\n",
       "      <td>97721650</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>1</td>\n",
       "      <td>97721542</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>1</td>\n",
       "      <td>97699535</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>1</td>\n",
       "      <td>97699533</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>52</th>\n",
       "      <td>1</td>\n",
       "      <td>97699506</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>1</td>\n",
       "      <td>97699474</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>1</td>\n",
       "      <td>97699430</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>1</td>\n",
       "      <td>97699399</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>1</td>\n",
       "      <td>97679170</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>1</td>\n",
       "      <td>97595149</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>1</td>\n",
       "      <td>97595088</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>1</td>\n",
       "      <td>97595083</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr variant_start reference_allele variant_allele  count\n",
       "148  19      38500898                C              G      2\n",
       "135  19      38496283                C              T      2\n",
       "124  19      38448712                G              C      2\n",
       "122  19      38446710                G              A      2\n",
       "213   6      18130687                T              G      2\n",
       "362   X     154536019                G              A      2\n",
       "363   X     154536019                G              T      2\n",
       "348   X     154535270                A              C      2\n",
       "349   X     154535270                A              G      2\n",
       "326   X     154534389                C              G      2\n",
       "327   X     154534389                C              T      2\n",
       "306   X     154533605                T              A      2\n",
       "307   X     154533605                T              C      2\n",
       "304   X     154533596                C              A      2\n",
       "305   X     154533596                C              G      2\n",
       "279   X     154532765                G              C      2\n",
       "280   X     154532765                G              T      2\n",
       "251   X     154532625                C              G      2\n",
       "252   X     154532625                C              T      2\n",
       "234   X     154532269                C              A      2\n",
       "235   X     154532269                C              G      2\n",
       "231   X     154532258                G              A      2\n",
       "232   X     154532258                G              T      2\n",
       "223   7      99672916                T              C      1\n",
       "222   7      99665212                C              T      1\n",
       "59    1      97883368                G              C      1\n",
       "58    1      97883353                G              A      1\n",
       "57    1      97883352                C              T      1\n",
       "56    1      97721650                T              C      1\n",
       "55    1      97721542                T              C      1\n",
       "54    1      97699535                T              C      1\n",
       "53    1      97699533                C              T      1\n",
       "52    1      97699506                C              T      1\n",
       "51    1      97699474                T              C      1\n",
       "50    1      97699430                T              G      1\n",
       "49    1      97699399                T              C      1\n",
       "48    1      97679170                T              C      1\n",
       "47    1      97595149                T              C      1\n",
       "46    1      97595088                A              G      1\n",
       "45    1      97595083                G              A      1"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# 用于人工校验重复的变异\n",
    "__df = df_variant.groupby([\"chr\", \"variant_start\", \"reference_allele\", \"variant_allele\"])[\"haplotype_name\"].count().reset_index(name='count')\n",
    "__df.sort_values(by=[\"count\", \"variant_start\"], ascending=False)[:40]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "fc5a6097",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_variant.to_csv(\"output/dataset_clean_snp.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "b2d7b8b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_variant = pd.read_csv(\"output/dataset_clean_snp_review.csv\", dtype=str).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "9b9fa03c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "377"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_variant)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "300fcd70",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>chr</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CYP2B6</td>\n",
       "      <td>19</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CYP2C19</td>\n",
       "      <td>10</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CYP2C9</td>\n",
       "      <td>10</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>22</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CYP3A5</td>\n",
       "      <td>7</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>DPYD</td>\n",
       "      <td>1</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>G6PD</td>\n",
       "      <td>X</td>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>NUDT15</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>RYR1</td>\n",
       "      <td>19</td>\n",
       "      <td>44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SLCO1B1</td>\n",
       "      <td>12</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>TPMT</td>\n",
       "      <td>6</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>UGT1A1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene chr  count\n",
       "0    CYP2B6  19     26\n",
       "1   CYP2C19  10     19\n",
       "2    CYP2C9  10     33\n",
       "3    CYP2D6  22     27\n",
       "4    CYP3A5   7      2\n",
       "5      DPYD   1     60\n",
       "6      G6PD   X    148\n",
       "7    NUDT15  13      1\n",
       "8      RYR1  19     44\n",
       "9   SLCO1B1  12      5\n",
       "10     TPMT   6     10\n",
       "11   UGT1A1   2      2"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## 药物基因上的变异数\n",
    "df_variant.groupby([\"gene\", \"chr\"])[\"haplotype_name\"].count().reset_index(name='count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "a6a32d25",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>function</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>decreased function</td>\n",
       "      <td>75</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>increased function</td>\n",
       "      <td>45</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>no function</td>\n",
       "      <td>192</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>normal function</td>\n",
       "      <td>65</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             function  count\n",
       "0  decreased function     75\n",
       "1  increased function     45\n",
       "2         no function    192\n",
       "3     normal function     65"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## 不同功能分组的变异数\n",
    "df_variant.groupby([\"function\"])[\"haplotype_name\"].count().reset_index(name='count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "d3108cbe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['1', '10', '12', '13', '19', '2', '22', '6', '7', 'X']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## 涉及到的染色体\n",
    "sorted(set(df_variant[\"chr\"].values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "f5e22a45",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr annotation search\n",
    "# 对位点排序加速查询，n^2 时间复杂度提升至n*log(n)\n",
    "\n",
    "def write_variant_file(df_chr, chr_path, output_path, chr_num):\n",
    "    def gen_file(total_str, variant_start, filepath, first=True):\n",
    "        variant_len = len(variant_start)\n",
    "        count = 0\n",
    "        cur_str = \"{}\t{}\".format(chr_num, variant_start[count])\n",
    "        matched = False\n",
    "\n",
    "        all_pos_list = [\"{}\t\".format(chr_num) + x for x in variant_start]\n",
    "        matched_pos_list = []\n",
    "\n",
    "        with open(chr_path, \"r\") as f:\n",
    "            if first:\n",
    "                total_str += f.readline()\n",
    "\n",
    "            line = f.readline()\n",
    "            while line:\n",
    "                if matched and not line.startswith(cur_str):\n",
    "                    count += 1\n",
    "                    if count >= variant_len:\n",
    "                        break\n",
    "                    matched = False\n",
    "                    cur_str = \"{}\t{}\".format(chr_num, variant_start[count])\n",
    "\n",
    "                if line.startswith(cur_str) and count < variant_len:\n",
    "\n",
    "                    matched_pos_list.append(cur_str)\n",
    "                    matched = True\n",
    "                    total_str += line\n",
    "\n",
    "                line = f.readline()\n",
    "\n",
    "        matched_pos_list = list(set(matched_pos_list))\n",
    "        missed_pos_list = sorted(set(all_pos_list) - set(matched_pos_list))\n",
    "        return missed_pos_list, total_str\n",
    "\n",
    "\n",
    "    variant_list = list(df_chr[\"variant_start\"].values)\n",
    "    variant_str = \"\"\n",
    "\n",
    "    missed_pos_list, variant_str = gen_file(variant_str, variant_list, chr_path)\n",
    "    print(missed_pos_list)\n",
    "\n",
    "    while len(missed_pos_list) > 0:\n",
    "        missed_pos_list.pop(0)\n",
    "        \n",
    "        if len(missed_pos_list) == 0:\n",
    "            break\n",
    "            \n",
    "        missed_pos_list = [x.split(\"\\t\")[-1] for x in missed_pos_list]\n",
    "        missed_pos_list, variant_str = gen_file(variant_str, missed_pos_list, chr_path, first=False)\n",
    "        print(missed_pos_list)\n",
    "        \n",
    "    with open(output_path, \"w\") as f:\n",
    "        f.write(variant_str)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "827762bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr1 DPYD\n",
    "df_chr1 = df_variant[df_variant[\"chr\"] == \"1\"]\n",
    "df_chr1 = df_chr1.sort_values(by=[\"variant_start\"])\n",
    "path = \"d:/download/dbNSFP4.2a_variant.chr1\"\n",
    "output_file = \"output/chr1_variant.tsv\"\n",
    "\n",
    "# write_variant_file(df_chr1, path, output_file, \"1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "cb1eda8a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr10 CYP2C19 CYP2C9\n",
    "df_chr10 = df_variant[df_variant[\"chr\"] == \"10\"]\n",
    "df_chr10 = df_chr10.sort_values(by=[\"variant_start\"])\n",
    "path = \"d:/download/dbNSFP4.2a_variant.chr10\"\n",
    "output_file = \"output/chr10_variant.tsv\"\n",
    "\n",
    "# write_variant_file(df_chr10, path, output_file, \"10\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "9ef46849",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr19 RYR1\n",
    "df_chr19 = df_variant[df_variant[\"chr\"] == \"19\"]\n",
    "df_chr19 = df_chr19.sort_values(by=[\"variant_start\"])\n",
    "path = \"d:/download/dbNSFP4.2a_variant.chr19\"\n",
    "output_file = \"output/chr19_variant.tsv\"\n",
    "\n",
    "# write_variant_file(df_chr19, path, output_file, \"19\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "1e3c4d50",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr22 CYP2D6\n",
    "df_chr22 = df_variant[df_variant[\"chr\"] == \"22\"]\n",
    "df_chr22 = df_chr22.sort_values(by=[\"variant_start\"])\n",
    "path = \"d:/download/dbNSFP4.2a_variant.chr22\"\n",
    "output_file = \"output/chr22_variant.tsv\"\n",
    "\n",
    "# write_variant_file(df_chr22, path, output_file, \"22\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "a435adb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr12 SLCO1B1\n",
    "df_chr12 = df_variant[df_variant[\"chr\"] == \"12\"]\n",
    "df_chr12 = df_chr12.sort_values(by=[\"variant_start\"])\n",
    "path = \"d:/download/dbNSFP4.2a_variant.chr12\"\n",
    "output_file = \"output/chr12_variant.tsv\"\n",
    "\n",
    "# write_variant_file(df_chr12, path, output_file, \"12\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "07bd2d6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr13 NUDT15\n",
    "df_chr13 = df_variant[df_variant[\"chr\"] == \"13\"]\n",
    "df_chr13 = df_chr13.sort_values(by=[\"variant_start\"])\n",
    "path = \"d:/download/dbNSFP4.2a_variant.chr13\"\n",
    "output_file = \"output/chr13_variant.tsv\"\n",
    "\n",
    "# write_variant_file(df_chr13, path, output_file, \"13\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "706b4cf2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#chr2 UGT1A1\n",
    "df_chr2 = df_variant[df_variant[\"chr\"] == \"2\"]\n",
    "df_chr2 = df_chr2.sort_values(by=[\"variant_start\"])\n",
    "path = \"d:/download/dbNSFP4.2a_variant.chr2\"\n",
    "output_file = \"output/chr2_variant.tsv\"\n",
    "\n",
    "# write_variant_file(df_chr2, path, output_file, \"2\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "700c826a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#chr6 TPMT\n",
    "df_chr6 = df_variant[df_variant[\"chr\"] == \"6\"]\n",
    "df_chr6 = df_chr6.sort_values(by=[\"variant_start\"])\n",
    "path = \"d:/download/dbNSFP4.2a_variant.chr6\"\n",
    "output_file = \"output/chr6_variant.tsv\"\n",
    "\n",
    "# write_variant_file(df_chr6, path, output_file, \"6\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "7608063d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#chr7 CYP3A5\n",
    "df_chr7 = df_variant[df_variant[\"chr\"] == \"7\"]\n",
    "df_chr7 = df_chr7.sort_values(by=[\"variant_start\"])\n",
    "path = \"d:/download/dbNSFP4.2a_variant.chr7\"\n",
    "output_file = \"output/chr7_variant.tsv\"\n",
    "\n",
    "# write_variant_file(df_chr7, path, output_file, \"7\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "b872d602",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['X\\t154532278', 'X\\t154532279', 'X\\t154532389', 'X\\t154532390', 'X\\t154532392', 'X\\t154532403', 'X\\t154532408', 'X\\t154532411', 'X\\t154532432', 'X\\t154532434', 'X\\t154532458', 'X\\t154532459', 'X\\t154532570', 'X\\t154532608', 'X\\t154532623', 'X\\t154532625', 'X\\t154532626', 'X\\t154532628', 'X\\t154532629', 'X\\t154532634', 'X\\t154532639', 'X\\t154532649', 'X\\t154532661', 'X\\t154532662', 'X\\t154532667', 'X\\t154532674', 'X\\t154532676', 'X\\t154532677', 'X\\t154532679', 'X\\t154532688', 'X\\t154532692', 'X\\t154532694', 'X\\t154532695', 'X\\t154532698', 'X\\t154532699', 'X\\t154532700', 'X\\t154532701', 'X\\t154532713', 'X\\t154532715', 'X\\t154532716', 'X\\t154532722', 'X\\t154532758', 'X\\t154532765', 'X\\t154532772', 'X\\t154532773', 'X\\t154532797', 'X\\t154532802', 'X\\t154532945', 'X\\t154532969', 'X\\t154532987', 'X\\t154532989', 'X\\t154532990', 'X\\t154533004', 'X\\t154533016', 'X\\t154533029', 'X\\t154533031', 'X\\t154533044', 'X\\t154533064', 'X\\t154533072', 'X\\t154533077', 'X\\t154533083', 'X\\t154533122', 'X\\t154533586', 'X\\t154533587', 'X\\t154533589', 'X\\t154533592', 'X\\t154533596', 'X\\t154533605', 'X\\t154533607', 'X\\t154533608', 'X\\t154533614', 'X\\t154533615', 'X\\t154533619', 'X\\t154533620', 'X\\t154533629', 'X\\t154533634', 'X\\t154534036', 'X\\t154534092', 'X\\t154534102', 'X\\t154534110', 'X\\t154534125', 'X\\t154534126', 'X\\t154534157', 'X\\t154534345', 'X\\t154534348', 'X\\t154534389', 'X\\t154534390', 'X\\t154534409', 'X\\t154534419', 'X\\t154534440', 'X\\t154534447', 'X\\t154534455', 'X\\t154534463', 'X\\t154534465', 'X\\t154534468', 'X\\t154534485', 'X\\t154534486', 'X\\t154534489', 'X\\t154534494', 'X\\t154534495', 'X\\t154535180', 'X\\t154535187', 'X\\t154535211', 'X\\t154535244', 'X\\t154535247', 'X\\t154535261', 'X\\t154535270', 'X\\t154535277', 'X\\t154535301', 'X\\t154535316', 'X\\t154535330', 'X\\t154535379', 'X\\t154535962', 'X\\t154535963', 'X\\t154535980', 'X\\t154535995', 'X\\t154535996', 'X\\t154536002', 'X\\t154536008', 'X\\t154536025', 'X\\t154536032', 'X\\t154536034', 'X\\t154536035', 'X\\t154536045', 'X\\t154536151', 'X\\t154536156', 'X\\t154536168', 'X\\t154536169', 'X\\t154546061', 'X\\t154546116', 'X\\t154546122']\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "#chrX D6PD\n",
    "df_chrX = df_variant[df_variant[\"chr\"] == \"X\"]\n",
    "df_chrX = df_chrX.sort_values(by=[\"variant_start\"])\n",
    "path = \"d:/download/dbNSFP4.2a_variant.chrX\"\n",
    "output_file = \"output/chrX_variant.tsv\"\n",
    "\n",
    "write_variant_file(df_chrX, path, output_file, \"X\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d62d13ff",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "6d56d771",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_var = pd.read_csv(\"output/chr6_variant.tsv\", sep=\"\\t\", dtype=str).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bc539957",
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_variant_annotation(df_variant, df_annotation, remain_column):\n",
    "    df_anno = pd.DataFrame(columns=remain_column)\n",
    "    exist_list = []\n",
    "    for index, row in df_variant.iterrows():\n",
    "        df_annotation_new = df_annotation[\n",
    "            (df_annotation[\"pos(1-based)\"] == row[\"variant_start\"]) &\n",
    "            (df_annotation[\"ref\"] == row[\"reference_allele\"]) &\n",
    "            (df_annotation[\"alt\"] == row[\"variant_allele\"])\n",
    "        ]\n",
    "        \n",
    "        df_annotation_new = df_annotation_new[remain_column]\n",
    "        df_annotation_new.index = range(len(df_annotation_new))\n",
    "        if len(df_annotation_new) == 1:\n",
    "            df_anno = pd.concat([df_anno, df_annotation_new], axis=0)\n",
    "            exist_list.append(True)\n",
    "        elif len(df_annotation_new) == 0:\n",
    "            df_anno = pd.concat([df_anno, pd.DataFrame([[\"\"] * len(remain_column)], columns=remain_column)], axis=0)\n",
    "            exist_list.append(False)\n",
    "        else:\n",
    "            df_anno = pd.concat([df_anno, df_annotation_new.iloc[:1]], axis=0)\n",
    "            exist_list.append(True)\n",
    "            \n",
    "    df_anno.index = range(len(df_anno))\n",
    "    df_variant.index = range(len(df_variant))\n",
    "    df_variant = pd.concat([df_variant, df_anno], axis=1)\n",
    "    df_variant[\"exist\"] = exist_list\n",
    "    df_variant = df_variant[df_variant[\"exist\"] == True]\n",
    "    df_variant.pop(\"exist\")\n",
    "    return df_variant\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f3cedf0",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4fa61a32",
   "metadata": {},
   "outputs": [],
   "source": [
    "score_rank_list = [\n",
    "    \"DEOGEN2_score\", \"MPC_score\", \"integrated_fitCons_score\", \"FATHMM_score\", \"LIST-S2_score\",\n",
    "    \"PrimateAI_score\", \"PROVEAN_score\", \"MutationAssessor_score\", \"BayesDel_addAF_score\", \n",
    "    \"MetaLR_score\", \"MetaSVM_score\", \"VEST4_score\", \"BayesDel_noAF_score\", \"GERP++_RS\", \n",
    "    \"GenoCanyon_score\", \"ClinPred_score\", \"MVP_score\", \"REVEL_score\", \"SIFT_score\", \"Polyphen2_HDIV_score\",\n",
    "    \"Polyphen2_HVAR_score\", \"phastCons17way_primate\", \"bStatistic\", \"DANN_score\", \"CADD_raw\"\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "68819540",
   "metadata": {},
   "outputs": [],
   "source": [
    "score_column = list(filter(lambda x: \"score\" in x or \"primate\" in x or \"prob\" in x, df_var.columns))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c416424f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "selected_column = list(set(score_rank_list).union(set(score_column)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "af0a8643",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame({\"column\": selected_column}).to_csv(\"output/selected_column.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "5b8a077e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr1 feature\n",
    "df_chr1_anno = pd.read_csv(\"output/chr1_variant.tsv\", sep=\"\\t\", dtype=str).fillna(\"\")\n",
    "df_chr1_feat = generate_variant_annotation(df_chr1, df_chr1_anno, selected_column)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "fa53cde7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr10 feature\n",
    "df_chr10_anno = pd.read_csv(\"output/chr10_variant.tsv\", sep=\"\\t\", dtype=str).fillna(\"\")\n",
    "df_chr10_feat = generate_variant_annotation(df_chr10, df_chr10_anno, selected_column)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "5ef994c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr12 feature\n",
    "df_chr12_anno = pd.read_csv(\"output/chr12_variant.tsv\", sep=\"\\t\", dtype=str).fillna(\"\")\n",
    "df_chr12_feat = generate_variant_annotation(df_chr12, df_chr12_anno, selected_column)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "7a57ad8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr13 feature\n",
    "df_chr13_anno = pd.read_csv(\"output/chr13_variant.tsv\", sep=\"\\t\", dtype=str).fillna(\"\")\n",
    "df_chr13_feat = generate_variant_annotation(df_chr13, df_chr13_anno, selected_column)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "2bd813df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr19 feature\n",
    "df_chr19_anno = pd.read_csv(\"output/chr19_variant.tsv\", sep=\"\\t\", dtype=str).fillna(\"\")\n",
    "df_chr19_feat = generate_variant_annotation(df_chr19, df_chr19_anno, selected_column)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "47ce08de",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr2 feature\n",
    "df_chr2_anno = pd.read_csv(\"output/chr2_variant.tsv\", sep=\"\\t\", dtype=str).fillna(\"\")\n",
    "df_chr2_feat = generate_variant_annotation(df_chr2, df_chr2_anno, selected_column)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "30507362",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr22 feature\n",
    "df_chr22_anno = pd.read_csv(\"output/chr22_variant.tsv\", sep=\"\\t\", dtype=str).fillna(\"\")\n",
    "df_chr22_feat = generate_variant_annotation(df_chr22, df_chr22_anno, selected_column)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "e95677dc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr6 feature\n",
    "df_chr6_anno = pd.read_csv(\"output/chr6_variant.tsv\", sep=\"\\t\", dtype=str).fillna(\"\")\n",
    "df_chr6_feat = generate_variant_annotation(df_chr6, df_chr6_anno, selected_column)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "e3cf38d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chr7 feature\n",
    "# no result\n",
    "df_chr7_anno = pd.read_csv(\"output/chr7_variant.tsv\", sep=\"\\t\", dtype=str).fillna(\"\")\n",
    "df_chr7_feat = generate_variant_annotation(df_chr7, df_chr7_anno, selected_column)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "f602a40e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# chrX feature\n",
    "df_chrX_anno = pd.read_csv(\"output/chrX_variant.tsv\", sep=\"\\t\", dtype=str).fillna(\"\")\n",
    "df_chrX_feat = generate_variant_annotation(df_chrX, df_chrX_anno, selected_column)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "0508ffbf",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_feat = pd.concat([df_chr1_feat, df_chr10_feat, df_chr12_feat, df_chr13_feat, \n",
    "                     df_chr19_feat, df_chr2_feat, df_chr22_feat, df_chr6_feat, \n",
    "                     df_chr7_feat, df_chrX_feat], axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3f0190a",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "901202a1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "_df1 = df_feat.groupby([\"gene\", \"chr\"])[\"haplotype_name\"].count().reset_index(name='gene_count')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "06d8e471",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>function</th>\n",
       "      <th>count</th>\n",
       "      <th>percentage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CYP2B6</td>\n",
       "      <td>decreased function</td>\n",
       "      <td>4</td>\n",
       "      <td>16.0%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CYP2B6</td>\n",
       "      <td>increased function</td>\n",
       "      <td>1</td>\n",
       "      <td>4.0%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CYP2B6</td>\n",
       "      <td>no function</td>\n",
       "      <td>12</td>\n",
       "      <td>48.0%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CYP2B6</td>\n",
       "      <td>normal function</td>\n",
       "      <td>8</td>\n",
       "      <td>32.0%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CYP2C19</td>\n",
       "      <td>decreased function</td>\n",
       "      <td>6</td>\n",
       "      <td>33.33%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>CYP2C19</td>\n",
       "      <td>no function</td>\n",
       "      <td>8</td>\n",
       "      <td>44.44%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>CYP2C19</td>\n",
       "      <td>normal function</td>\n",
       "      <td>4</td>\n",
       "      <td>22.22%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CYP2C9</td>\n",
       "      <td>decreased function</td>\n",
       "      <td>21</td>\n",
       "      <td>63.64%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CYP2C9</td>\n",
       "      <td>no function</td>\n",
       "      <td>11</td>\n",
       "      <td>33.33%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CYP2C9</td>\n",
       "      <td>normal function</td>\n",
       "      <td>1</td>\n",
       "      <td>3.03%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>decreased function</td>\n",
       "      <td>3</td>\n",
       "      <td>11.54%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>no function</td>\n",
       "      <td>18</td>\n",
       "      <td>69.23%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CYP2D6</td>\n",
       "      <td>normal function</td>\n",
       "      <td>5</td>\n",
       "      <td>19.23%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>DPYD</td>\n",
       "      <td>decreased function</td>\n",
       "      <td>4</td>\n",
       "      <td>7.02%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>DPYD</td>\n",
       "      <td>no function</td>\n",
       "      <td>14</td>\n",
       "      <td>24.56%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>DPYD</td>\n",
       "      <td>normal function</td>\n",
       "      <td>39</td>\n",
       "      <td>68.42%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>G6PD</td>\n",
       "      <td>decreased function</td>\n",
       "      <td>33</td>\n",
       "      <td>22.45%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>G6PD</td>\n",
       "      <td>no function</td>\n",
       "      <td>110</td>\n",
       "      <td>74.83%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>G6PD</td>\n",
       "      <td>normal function</td>\n",
       "      <td>4</td>\n",
       "      <td>2.72%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NUDT15</td>\n",
       "      <td>no function</td>\n",
       "      <td>1</td>\n",
       "      <td>100.0%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>RYR1</td>\n",
       "      <td>decreased function</td>\n",
       "      <td>1</td>\n",
       "      <td>2.33%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>RYR1</td>\n",
       "      <td>increased function</td>\n",
       "      <td>42</td>\n",
       "      <td>97.67%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>SLCO1B1</td>\n",
       "      <td>no function</td>\n",
       "      <td>4</td>\n",
       "      <td>80.0%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SLCO1B1</td>\n",
       "      <td>normal function</td>\n",
       "      <td>1</td>\n",
       "      <td>20.0%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>TPMT</td>\n",
       "      <td>no function</td>\n",
       "      <td>9</td>\n",
       "      <td>100.0%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>UGT1A1</td>\n",
       "      <td>decreased function</td>\n",
       "      <td>2</td>\n",
       "      <td>100.0%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       gene            function  count percentage\n",
       "0    CYP2B6  decreased function      4      16.0%\n",
       "1    CYP2B6  increased function      1       4.0%\n",
       "2    CYP2B6         no function     12      48.0%\n",
       "3    CYP2B6     normal function      8      32.0%\n",
       "4   CYP2C19  decreased function      6     33.33%\n",
       "5   CYP2C19         no function      8     44.44%\n",
       "6   CYP2C19     normal function      4     22.22%\n",
       "7    CYP2C9  decreased function     21     63.64%\n",
       "8    CYP2C9         no function     11     33.33%\n",
       "9    CYP2C9     normal function      1      3.03%\n",
       "10   CYP2D6  decreased function      3     11.54%\n",
       "11   CYP2D6         no function     18     69.23%\n",
       "12   CYP2D6     normal function      5     19.23%\n",
       "13     DPYD  decreased function      4      7.02%\n",
       "14     DPYD         no function     14     24.56%\n",
       "15     DPYD     normal function     39     68.42%\n",
       "16     G6PD  decreased function     33     22.45%\n",
       "17     G6PD         no function    110     74.83%\n",
       "18     G6PD     normal function      4      2.72%\n",
       "19   NUDT15         no function      1     100.0%\n",
       "20     RYR1  decreased function      1      2.33%\n",
       "21     RYR1  increased function     42     97.67%\n",
       "22  SLCO1B1         no function      4      80.0%\n",
       "23  SLCO1B1     normal function      1      20.0%\n",
       "24     TPMT         no function      9     100.0%\n",
       "25   UGT1A1  decreased function      2     100.0%"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_df2 = df_feat.groupby([\"gene\", \"function\"])[\"haplotype_name\"].count().reset_index(name='count')\n",
    "_df2 = pd.merge(_df2, _df1, how='left', on=['gene'])\n",
    "_df2 = _df2.assign(percentage=_df2.apply(lambda row: \"{}%\".format(round(row['count'] / row['gene_count'] * 100, 2)), axis=1))\n",
    "_df2.pop(\"chr\")\n",
    "_df2.pop(\"gene_count\")\n",
    "_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "b7e931cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>function</th>\n",
       "      <th>count</th>\n",
       "      <th>percentage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>decreased function</td>\n",
       "      <td>74</td>\n",
       "      <td>20.22%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>increased function</td>\n",
       "      <td>43</td>\n",
       "      <td>11.75%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>no function</td>\n",
       "      <td>187</td>\n",
       "      <td>51.09%</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>normal function</td>\n",
       "      <td>62</td>\n",
       "      <td>16.94%</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             function  count percentage\n",
       "0  decreased function     74     20.22%\n",
       "1  increased function     43     11.75%\n",
       "2         no function    187     51.09%\n",
       "3     normal function     62     16.94%"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "_df2 = df_feat.groupby([\"function\"])[\"haplotype_name\"].count().reset_index(name='count')\n",
    "sum_count = sum(_df2[\"count\"].values)\n",
    "_df2 = _df2.assign(percentage=_df2.apply(lambda row: \"{}%\".format(round(row['count'] / sum_count * 100, 2)), axis=1))\n",
    "_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "c0c877dd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "366"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_feat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "e4d16d02",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_feat.to_csv(\"output/variant_feat.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "c1c7d485",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "86"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(selected_column)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "1391e415",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_lof = pd.read_csv(\"data/LoFtool_scores.txt\", sep=\"\\t\")\n",
    "lof_dict = dict(zip(\n",
    "    df_lof[\"Gene\"].values,\n",
    "    df_lof[\"LoFtool_percentile\"].values\n",
    "))\n",
    "gene_lof_list = [str(lof_dict.get(x, \"\")) for x in df_feat[\"gene\"].values]\n",
    "\n",
    "df_feat[\"LoFtool\"] = gene_lof_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "9bba5e1b",
   "metadata": {},
   "outputs": [],
   "source": [
    "chr_list = df_feat[\"chr\"].values\n",
    "variant_start_list = df_feat[\"variant_start\"].values\n",
    "reference_allele_list = df_feat[\"reference_allele\"].values\n",
    "variant_allele_list = df_feat[\"variant_allele\"].values\n",
    "\n",
    "feat_variant = [\n",
    "    \"{}:g.{}{}>{}\".format(x[0], x[1], x[2], x[3]) for x in \n",
    "     zip(chr_list, variant_start_list, reference_allele_list, variant_allele_list)\n",
    "]\n",
    "df_feat[\"variant\"] = feat_variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "28aae1a5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# clean_feat\n",
    "\n",
    "def is_float(value):\n",
    "    try:\n",
    "        float(value)\n",
    "        return True\n",
    "    except:\n",
    "        return False\n",
    "\n",
    "\n",
    "for index, row in df_feat.iterrows():\n",
    "    for col in df_feat.columns:\n",
    "        if row[col] == '.':\n",
    "            row[col] = \"\"\n",
    "\n",
    "        if \";\" in row[col]:\n",
    "            val_str = row[col].strip().split(\";\")\n",
    "            val_str = list(filter(lambda x: x!= \"\" and x!= \".\" and is_float(x), val_str))\n",
    "            if len(val_str) > 0:\n",
    "                row[col] = str(min([float(x) for x in val_str]))\n",
    "            else:\n",
    "                row[col] = \"\"\n",
    "\n",
    "\n",
    "df_feat.to_csv(\"output/variant_feat_clean.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "10c2f5fe",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "aca05f03",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import pandas as pd\n",
    "import os\n",
    "import json\n",
    "\n",
    "os.chdir(\"D:/functional-prediction/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "3dd3912c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_snp = pd.read_csv(\"output/dataset_clean_snp_review.csv\", dtype=str).fillna(\"\")\n",
    "\n",
    "chr_list = df_snp[\"chr\"].values\n",
    "variant_start_list = df_snp[\"variant_start\"].values\n",
    "reference_allele_list = df_snp[\"reference_allele\"].values\n",
    "variant_allele_list = df_snp[\"variant_allele\"].values\n",
    "\n",
    "notations = [\n",
    "    \"{}:g.{}{}>{}\".format(x[0], x[1], x[2], x[3]) for x in \n",
    "     zip(chr_list, variant_start_list, reference_allele_list, variant_allele_list)\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "f2fa103c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# VEP data source\n",
    "# https://rest.ensembl.org/documentation/info/vep_hgvs_get\n",
    "\n",
    "path = \"https://rest.ensembl.org/vep/human/hgvs/\"\n",
    "headers={ \"Content-Type\" : \"application/json\", \"Accept\" : \"application/json\"}\n",
    "\n",
    "batch = 100\n",
    "\n",
    "notation_list = []\n",
    "\n",
    "for i in range(int(len(notations) / batch) + 1):\n",
    "    print(i)\n",
    "    r = requests.post(\n",
    "        path,\n",
    "        headers=headers,\n",
    "        json={ \n",
    "            \"hgvs_notations\" : notations[i * batch : (i + 1) * batch]\n",
    "        },\n",
    "        timeout=30\n",
    "    )\n",
    "    resp_list = json.loads(r.text)\n",
    "    notation_list.extend(resp_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "9ed7bb4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "notation_dict = {x[\"id\"]: x for x in notation_list}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "532dfda1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "52d448d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"output/notation_dict.json\", \"w\") as f:\n",
    "    json.dump(notation_dict, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "77ec1a79",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('missense_variant', 345),\n",
       " ('stop_gained', 15),\n",
       " ('synonymous_variant', 6),\n",
       " ('splice_acceptor_variant', 5),\n",
       " ('start_lost', 3),\n",
       " ('splice_donor_variant', 2),\n",
       " ('upstream_gene_variant', 1)]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from collections import Counter\n",
    "Counter([x[\"most_severe_consequence\"] for x in notation_list]).most_common()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "7218b0e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "dict_keys(['id', 'colocated_variants', 'allele_string', 'assembly_name', 'seq_region_name', 'transcript_consequences', 'most_severe_consequence', 'start', 'end', 'input', 'strand'])"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "notation_list[1].keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "64bc7a7f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 通过变异位点获取基因名\n",
    "gene_symbol_list = []\n",
    "for notat in notation_list:\n",
    "    gene_symbol_list.append(list(set([x.get(\"gene_symbol\", \"\") for x in notat[\"transcript_consequences\"]]) - {\"\"})[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "3db2a1f2",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_lof = pd.read_csv(\"data/LoFtool_scores.txt\", sep=\"\\t\")\n",
    "lof_dict = dict(zip(\n",
    "    df_lof[\"Gene\"].values,\n",
    "    df_lof[\"LoFtool_percentile\"].values\n",
    "))\n",
    "gene_lof_list = [lof_dict.get(x, \"\") for x in gene_symbol_list]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5a61db8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "e4b1ad09",
   "metadata": {},
   "outputs": [],
   "source": [
    "notat_dict = {}\n",
    "all_notations = []\n",
    "for x in notation_list:\n",
    "    ct_list = []\n",
    "    for y in x[\"transcript_consequences\"]:\n",
    "        ct_list.extend(y[\"consequence_terms\"])\n",
    "    ct_list = list(set(ct_list))\n",
    "    all_notations.extend(ct_list)\n",
    "    notat_dict[x[\"id\"]] = {\"consequence_terms\": ct_list, \"most_severe_consequence\": x[\"most_severe_consequence\"]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d0292a7f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "0db57f5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "not_df_dict = {}\n",
    "\n",
    "not_df_dict[\"variant\"] = []\n",
    "\n",
    "for key in set(all_notations):\n",
    "    not_df_dict[key] = []\n",
    "    \n",
    "for key, value in notat_dict.items():\n",
    "    exist_cons = value[\"consequence_terms\"]\n",
    "    not_exist_cons = list(set(all_notations) - set(exist_cons))\n",
    "    for ec in exist_cons:\n",
    "        not_df_dict[ec].append(1)\n",
    "    for ec in not_exist_cons:\n",
    "        not_df_dict[ec].append(0)\n",
    "    not_df_dict[\"variant\"].append(key)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "f8d3967d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_notation = pd.DataFrame(not_df_dict)\n",
    "df_notation.to_csv(\"output/snp_variant_type.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "5eb15e2d",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "upstream_gene_variant: 143\n",
      "synonymous_variant: 27\n",
      "NMD_transcript_variant: 273\n",
      "non_coding_transcript_variant: 38\n",
      "splice_donor_variant: 2\n",
      "splice_acceptor_variant: 5\n",
      "5_prime_UTR_variant: 4\n",
      "incomplete_terminal_codon_variant: 1\n",
      "non_coding_transcript_exon_variant: 214\n",
      "splice_region_variant: 14\n",
      "stop_gained: 15\n",
      "splice_polypyrimidine_tract_variant: 1\n",
      "start_lost: 3\n",
      "intron_variant: 72\n",
      "missense_variant: 348\n",
      "downstream_gene_variant: 175\n",
      "coding_sequence_variant: 1\n",
      "3_prime_UTR_variant: 178\n"
     ]
    }
   ],
   "source": [
    "for col in list(set(list(df_notation.columns)) - set(\"variant\")):\n",
    "    if col == \"variant\":\n",
    "        continue\n",
    "    print(\"{}: {}\".format(col, sum(df_notation[col].values)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2dea4a7d",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "3c1e3dd5",
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.merge(df_feat, df_notation, how=\"left\", on=[\"variant\"]).to_csv(\"output/variant_feat_clean.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18a8146b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "be99913b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 添加APF score\n",
    "# https://www.nature.com/articles/s41397-018-0044-2\n",
    "\n",
    "import yaml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "d3b3f7cc",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"threshold.yml\", \"r\") as f:\n",
    "    thres_dict = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "fe423bac",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_feat = pd.read_csv(\"output/variant_feat_clean.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "d516f375",
   "metadata": {},
   "outputs": [],
   "source": [
    "apf_score_list = []\n",
    "for index, row in df_feat.iterrows():\n",
    "    cur_list = []\n",
    "    for key, value in thres_dict.items():\n",
    "        score = row[key]\n",
    "        if not score:\n",
    "            continue\n",
    "        \n",
    "        if (value[\"type\"] == \"lt\" and score < value[\"threshold\"]) or \\\n",
    "           (value[\"type\"] == \"gt\" and score > value[\"threshold\"]):\n",
    "            cur_list.append(1)\n",
    "        else:\n",
    "            cur_list.append(0)\n",
    "        \n",
    "    if len(cur_list) == 0:\n",
    "        apf_score_list.append(None)\n",
    "        \n",
    "    else:\n",
    "        apf_score_list.append(round(sum(cur_list) / len(thres_dict.keys()), 3))\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "id": "d59827e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_feat[\"APF_score\"] = apf_score_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "ca9df380",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>haplotype_name</th>\n",
       "      <th>chr</th>\n",
       "      <th>variant_start</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>variant_allele</th>\n",
       "      <th>function</th>\n",
       "      <th>variant</th>\n",
       "      <th>type</th>\n",
       "      <th>Polyphen2_HDIV_score</th>\n",
       "      <th>...</th>\n",
       "      <th>splice_region_variant</th>\n",
       "      <th>stop_gained</th>\n",
       "      <th>splice_polypyrimidine_tract_variant</th>\n",
       "      <th>start_lost</th>\n",
       "      <th>intron_variant</th>\n",
       "      <th>missense_variant</th>\n",
       "      <th>downstream_gene_variant</th>\n",
       "      <th>coding_sequence_variant</th>\n",
       "      <th>3_prime_UTR_variant</th>\n",
       "      <th>APF_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DPYD</td>\n",
       "      <td>rs148799944</td>\n",
       "      <td>1</td>\n",
       "      <td>97078993</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>normal function</td>\n",
       "      <td>1:g.97078993C&gt;G</td>\n",
       "      <td>snp</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DPYD</td>\n",
       "      <td>rs140114515</td>\n",
       "      <td>1</td>\n",
       "      <td>97079005</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>normal function</td>\n",
       "      <td>1:g.97079005C&gt;T</td>\n",
       "      <td>snp</td>\n",
       "      <td>0.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DPYD</td>\n",
       "      <td>rs139459586</td>\n",
       "      <td>1</td>\n",
       "      <td>97079076</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>normal function</td>\n",
       "      <td>1:g.97079076A&gt;C</td>\n",
       "      <td>snp</td>\n",
       "      <td>0.954</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DPYD</td>\n",
       "      <td>rs202144771</td>\n",
       "      <td>1</td>\n",
       "      <td>97079077</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>normal function</td>\n",
       "      <td>1:g.97079077G&gt;A</td>\n",
       "      <td>snp</td>\n",
       "      <td>0.021</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DPYD</td>\n",
       "      <td>rs72547601</td>\n",
       "      <td>1</td>\n",
       "      <td>97079121</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>no function</td>\n",
       "      <td>1:g.97079121T&gt;C</td>\n",
       "      <td>snp</td>\n",
       "      <td>1.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.667</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>361</th>\n",
       "      <td>G6PD</td>\n",
       "      <td>G6PD Orissa</td>\n",
       "      <td>X</td>\n",
       "      <td>154536168</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>decreased function</td>\n",
       "      <td>X:g.154536168G&gt;C</td>\n",
       "      <td>snp</td>\n",
       "      <td>0.998</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362</th>\n",
       "      <td>G6PD</td>\n",
       "      <td>G6PD Rignano</td>\n",
       "      <td>X</td>\n",
       "      <td>154536169</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>decreased function</td>\n",
       "      <td>X:g.154536169C&gt;T</td>\n",
       "      <td>snp</td>\n",
       "      <td>1.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>363</th>\n",
       "      <td>G6PD</td>\n",
       "      <td>G6PD Gaohe</td>\n",
       "      <td>X</td>\n",
       "      <td>154546061</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>decreased function</td>\n",
       "      <td>X:g.154546061T&gt;C</td>\n",
       "      <td>snp</td>\n",
       "      <td>1.000</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>364</th>\n",
       "      <td>G6PD</td>\n",
       "      <td>G6PD Lages</td>\n",
       "      <td>X</td>\n",
       "      <td>154546116</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>decreased function</td>\n",
       "      <td>X:g.154546116C&gt;T</td>\n",
       "      <td>snp</td>\n",
       "      <td>0.999</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>365</th>\n",
       "      <td>G6PD</td>\n",
       "      <td>G6PD Sinnai</td>\n",
       "      <td>X</td>\n",
       "      <td>154546122</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>decreased function</td>\n",
       "      <td>X:g.154546122C&gt;A</td>\n",
       "      <td>snp</td>\n",
       "      <td>0.001</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>366 rows × 115 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene haplotype_name chr  variant_start reference_allele variant_allele  \\\n",
       "0    DPYD    rs148799944   1       97078993                C              G   \n",
       "1    DPYD    rs140114515   1       97079005                C              T   \n",
       "2    DPYD    rs139459586   1       97079076                A              C   \n",
       "3    DPYD    rs202144771   1       97079077                G              A   \n",
       "4    DPYD     rs72547601   1       97079121                T              C   \n",
       "..    ...            ...  ..            ...              ...            ...   \n",
       "361  G6PD    G6PD Orissa   X      154536168                G              C   \n",
       "362  G6PD   G6PD Rignano   X      154536169                C              T   \n",
       "363  G6PD     G6PD Gaohe   X      154546061                T              C   \n",
       "364  G6PD     G6PD Lages   X      154546116                C              T   \n",
       "365  G6PD    G6PD Sinnai   X      154546122                C              A   \n",
       "\n",
       "               function           variant type  Polyphen2_HDIV_score  ...  \\\n",
       "0       normal function   1:g.97078993C>G  snp                 0.000  ...   \n",
       "1       normal function   1:g.97079005C>T  snp                 0.000  ...   \n",
       "2       normal function   1:g.97079076A>C  snp                 0.954  ...   \n",
       "3       normal function   1:g.97079077G>A  snp                 0.021  ...   \n",
       "4           no function   1:g.97079121T>C  snp                 1.000  ...   \n",
       "..                  ...               ...  ...                   ...  ...   \n",
       "361  decreased function  X:g.154536168G>C  snp                 0.998  ...   \n",
       "362  decreased function  X:g.154536169C>T  snp                 1.000  ...   \n",
       "363  decreased function  X:g.154546061T>C  snp                 1.000  ...   \n",
       "364  decreased function  X:g.154546116C>T  snp                 0.999  ...   \n",
       "365  decreased function  X:g.154546122C>A  snp                 0.001  ...   \n",
       "\n",
       "     splice_region_variant  stop_gained  splice_polypyrimidine_tract_variant  \\\n",
       "0                        0            0                                    0   \n",
       "1                        0            0                                    0   \n",
       "2                        0            0                                    0   \n",
       "3                        0            0                                    0   \n",
       "4                        0            0                                    0   \n",
       "..                     ...          ...                                  ...   \n",
       "361                      0            0                                    0   \n",
       "362                      0            1                                    0   \n",
       "363                      0            0                                    0   \n",
       "364                      0            0                                    0   \n",
       "365                      0            0                                    0   \n",
       "\n",
       "     start_lost  intron_variant  missense_variant  downstream_gene_variant  \\\n",
       "0             0               0                 1                        0   \n",
       "1             0               0                 1                        0   \n",
       "2             0               0                 1                        0   \n",
       "3             0               0                 1                        0   \n",
       "4             0               0                 1                        0   \n",
       "..          ...             ...               ...                      ...   \n",
       "361           0               0                 1                        0   \n",
       "362           0               0                 1                        0   \n",
       "363           0               1                 1                        0   \n",
       "364           0               1                 1                        0   \n",
       "365           0               1                 1                        0   \n",
       "\n",
       "     coding_sequence_variant  3_prime_UTR_variant  APF_score  \n",
       "0                          0                    0      0.167  \n",
       "1                          0                    0      0.000  \n",
       "2                          0                    0      0.667  \n",
       "3                          0                    0      0.167  \n",
       "4                          0                    0      0.667  \n",
       "..                       ...                  ...        ...  \n",
       "361                        0                    0      0.500  \n",
       "362                        0                    0      0.333  \n",
       "363                        0                    0      0.833  \n",
       "364                        0                    0      0.833  \n",
       "365                        0                    0      0.500  \n",
       "\n",
       "[366 rows x 115 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_feat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "889af87c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_feat.to_csv(\"output/variant_feat_clean.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "106d428d",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(\"gene_range.yml\", \"r\") as f:\n",
    "    gene_range_dict = yaml.safe_load(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "ed5cf56f",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CYP2B6': {'chormosome': 19, 'start': 40991282, 'end': 41018398},\n",
       " 'CYP2C19': {'chormosome': 10, 'start': 94762681, 'end': 94855547},\n",
       " 'CYP2C9': {'chormosome': 10, 'start': 94938658, 'end': 94990091},\n",
       " 'CYP2D6': {'chormosome': 22, 'start': 42126499, 'end': 42130810},\n",
       " 'DPYD': {'chormosome': 1, 'start': 97077743, 'end': 97921059},\n",
       " 'NUDT15': {'chormosome': 13, 'start': 48037726, 'end': 48052755},\n",
       " 'RYR1': {'chormosome': 19, 'start': 38433691, 'end': 38587564},\n",
       " 'SLCO1B1': {'chormosome': 12, 'start': 21131194, 'end': 21239796},\n",
       " 'TPMT': {'chormosome': 6, 'start': 18128311, 'end': 18155169},\n",
       " 'UGT1A1': {'chormosome': 2, 'start': 233760270, 'end': 233773300}}"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_range_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "725feb21",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "df_indel = pd.read_csv(\"output/dataset_clean_indel.csv\", dtype=str).fillna(\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "id": "94dbf3b4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>type</th>\n",
       "      <th>function</th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>indel</td>\n",
       "      <td>decreased function</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>indel</td>\n",
       "      <td>no function</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    type            function  count\n",
       "0  indel  decreased function      1\n",
       "1  indel         no function     33"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_indel.groupby([\"type\", \"function\"])[\"haplotype_name\"].count().reset_index(name=\"count\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "79145bb7",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e75dcf5f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
