{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "f69625d9",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "os.chdir(\"D:/functional-prediction/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "5373867e",
   "metadata": {},
   "outputs": [],
   "source": [
    "match_dict = {\n",
    "    \"haplotype_name\": [],\n",
    "    \"variant\": [],\n",
    "    \"sign\": []\n",
    "}\n",
    "\n",
    "for fn in os.listdir(\"multi_var\"):\n",
    "    df_var = pd.read_csv(os.path.join(\"multi_var\", fn), dtype=str).fillna(\"\")\n",
    "    \n",
    "    for index, row in df_var.iterrows():\n",
    "        variant = row[\"variant\"]\n",
    "        for col in df_var.columns:\n",
    "            if col == \"variant\":\n",
    "                continue\n",
    "                \n",
    "            haplotype = col\n",
    "            sign = row[col]\n",
    "            if \"B1\" in sign:\n",
    "                match_dict[\"haplotype_name\"].append(haplotype)\n",
    "                match_dict[\"variant\"].append(variant)\n",
    "                match_dict[\"sign\"].append(sign)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "5ec25770",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_multi = pd.DataFrame(match_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "33d0d489",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "2b044a9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append(\"D:\\pgkb_graph\")\n",
    "os.chdir(\"D:\\pgkb_graph\")\n",
    "\n",
    "from src.util.variant_mapping_util import variantMappingUtil\n",
    "\n",
    "v = variantMappingUtil()\n",
    "\n",
    "function_dict = {}\n",
    "for gene, variant_dict in v.gene_functionality_dict.items():\n",
    "    for key, value in variant_dict.items():\n",
    "        if \"*\" in key:\n",
    "            function_dict[\"{}{}\".format(gene, key).strip()] = value[\"function_status\"]\n",
    "        else:\n",
    "            function_dict[\"{} {}\".format(gene, key).strip()] = value[\"function_status\"]\n",
    "            \n",
    "os.chdir(\"D:/functional-prediction/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "e132e8db",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_d = pd.read_csv(\"output/dataset.csv\", dtype=str).fillna(\"\")\n",
    "chr_dict = dict(zip(df_d[\"gene\"].values, df_d[\"chr\"].values))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "0d9f29ad",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "pos_regex = re.compile(r\"^[\\d]+\")\n",
    "\n",
    "info_dict = {\n",
    "    \"gene\": [],\n",
    "    \"chr\": [],\n",
    "    \"variant_start\": [],\n",
    "    \"reference_allele\": [],\n",
    "    \"variant_allele\": [],\n",
    "    \"function\": []\n",
    "}\n",
    "\n",
    "for index, row in df_multi.iterrows():\n",
    "    haplotype = row[\"haplotype_name\"]\n",
    "    function = function_dict.get(haplotype, \"\")\n",
    "    gene = haplotype.split(\"*\")[0]\n",
    "    chromo = chr_dict[gene]\n",
    "    variant = row[\"variant\"]\n",
    "    if \"ins\" in variant or \"del\" in variant:\n",
    "        variant_start = \"\"\n",
    "        reference_allele = \"\"\n",
    "        variant_allele = \"\"\n",
    "    else:\n",
    "        variant_start = re.findall(pos_regex, variant)[0]\n",
    "        variant = variant.replace(variant_start, \"\").strip()\n",
    "        reference_allele, variant_allele = variant.split(\">\")\n",
    "        \n",
    "    info_dict[\"gene\"].append(gene)\n",
    "    info_dict[\"chr\"].append(chromo)\n",
    "    info_dict[\"variant_start\"].append(variant_start)\n",
    "    info_dict[\"reference_allele\"].append(reference_allele)\n",
    "    info_dict[\"variant_allele\"].append(variant_allele)\n",
    "    info_dict[\"function\"].append(function)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "3a42a6fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_info = pd.DataFrame(info_dict)\n",
    "df_multi = pd.concat([df_multi, df_info], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "d077a46b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>haplotype_name</th>\n",
       "      <th>variant</th>\n",
       "      <th>sign</th>\n",
       "      <th>gene</th>\n",
       "      <th>chr</th>\n",
       "      <th>variant_start</th>\n",
       "      <th>reference_allele</th>\n",
       "      <th>variant_allele</th>\n",
       "      <th>function</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CYP2B6*10</td>\n",
       "      <td>40991367A&gt;T</td>\n",
       "      <td>B1</td>\n",
       "      <td>CYP2B6</td>\n",
       "      <td>19</td>\n",
       "      <td>40991367</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>Uncertain function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CYP2B6*17</td>\n",
       "      <td>40991381A&gt;T</td>\n",
       "      <td>B1</td>\n",
       "      <td>CYP2B6</td>\n",
       "      <td>19</td>\n",
       "      <td>40991381</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>Normal function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CYP2B6*17</td>\n",
       "      <td>40991388A&gt;G</td>\n",
       "      <td>B1</td>\n",
       "      <td>CYP2B6</td>\n",
       "      <td>19</td>\n",
       "      <td>40991388</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Normal function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CYP2B6*17</td>\n",
       "      <td>40991390C&gt;A</td>\n",
       "      <td>B1</td>\n",
       "      <td>CYP2B6</td>\n",
       "      <td>19</td>\n",
       "      <td>40991390</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>Normal function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CYP2B6*17</td>\n",
       "      <td>40991391G&gt;C</td>\n",
       "      <td>B1</td>\n",
       "      <td>CYP2B6</td>\n",
       "      <td>19</td>\n",
       "      <td>40991391</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Normal function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108</th>\n",
       "      <td>SLCO1B1*30</td>\n",
       "      <td>21178957A&gt;G</td>\n",
       "      <td>B1</td>\n",
       "      <td>SLCO1B1</td>\n",
       "      <td>12</td>\n",
       "      <td>21178957</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Uncertain function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>SLCO1B1*39</td>\n",
       "      <td>21196975C&gt;T</td>\n",
       "      <td>B1A</td>\n",
       "      <td>SLCO1B1</td>\n",
       "      <td>12</td>\n",
       "      <td>21196975</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Uncertain function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>110</th>\n",
       "      <td>SLCO1B1*47</td>\n",
       "      <td>21200544C&gt;G</td>\n",
       "      <td>B1</td>\n",
       "      <td>SLCO1B1</td>\n",
       "      <td>12</td>\n",
       "      <td>21200544</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>No function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>111</th>\n",
       "      <td>SLCO1B1*27</td>\n",
       "      <td>21202555C&gt;G</td>\n",
       "      <td>B1</td>\n",
       "      <td>SLCO1B1</td>\n",
       "      <td>12</td>\n",
       "      <td>21202555</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>Uncertain function</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>112</th>\n",
       "      <td>SLCO1B1*29</td>\n",
       "      <td>21239158C&gt;T</td>\n",
       "      <td>B1</td>\n",
       "      <td>SLCO1B1</td>\n",
       "      <td>12</td>\n",
       "      <td>21239158</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Uncertain function</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>113 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    haplotype_name      variant sign     gene chr variant_start  \\\n",
       "0        CYP2B6*10  40991367A>T   B1   CYP2B6  19      40991367   \n",
       "1        CYP2B6*17  40991381A>T   B1   CYP2B6  19      40991381   \n",
       "2        CYP2B6*17  40991388A>G   B1   CYP2B6  19      40991388   \n",
       "3        CYP2B6*17  40991390C>A   B1   CYP2B6  19      40991390   \n",
       "4        CYP2B6*17  40991391G>C   B1   CYP2B6  19      40991391   \n",
       "..             ...          ...  ...      ...  ..           ...   \n",
       "108     SLCO1B1*30  21178957A>G   B1  SLCO1B1  12      21178957   \n",
       "109     SLCO1B1*39  21196975C>T  B1A  SLCO1B1  12      21196975   \n",
       "110     SLCO1B1*47  21200544C>G   B1  SLCO1B1  12      21200544   \n",
       "111     SLCO1B1*27  21202555C>G   B1  SLCO1B1  12      21202555   \n",
       "112     SLCO1B1*29  21239158C>T   B1  SLCO1B1  12      21239158   \n",
       "\n",
       "    reference_allele variant_allele            function  \n",
       "0                  A              T  Uncertain function  \n",
       "1                  A              T     Normal function  \n",
       "2                  A              G     Normal function  \n",
       "3                  C              A     Normal function  \n",
       "4                  G              C     Normal function  \n",
       "..               ...            ...                 ...  \n",
       "108                A              G  Uncertain function  \n",
       "109                C              T  Uncertain function  \n",
       "110                C              G         No function  \n",
       "111                C              G  Uncertain function  \n",
       "112                C              T  Uncertain function  \n",
       "\n",
       "[113 rows x 9 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_multi"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "a1bde283",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_multi[\"function\"] = df_multi[\"function\"].str.lower().str.strip()\n",
    "\n",
    "df_multi = df_multi[(~df_multi[\"function\"].str.contains(\"uncertain\")) & (~df_multi[\"function\"].str.contains(\"unknown\"))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "421504f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir(\"D:/functional-prediction/\")\n",
    "df_multi[[\n",
    "    \"haplotype_name\", \"gene\", \"chr\", \"variant\", \n",
    "    \"variant_start\", \"reference_allele\", \n",
    "    \"variant_allele\", \"function\"\n",
    "]].to_csv(\"output/dataset_multi_var.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aad11947",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
